Introduce new_map_lit operation in the loader

michalmuskala opened the pull request:
Optimise construction of maps when all keys are literals but building a literal tuple in the loader. This benefits especially small maps since they basically consist of two tuples - one for the keys and one for the values.

This change should benefit especially Elixir, because of the record replacement - structs are always built with all literal keys. 

The change effectively means significantly reducing memory use (almost by a half) of programs using maps to replace records.

bjorng commented on erts/emulator/beam/beam_load.c:
```diff
@@ -5509,6 +5592,24 @@ new_literal(LoaderState* stp, Eterm** hpp, Uint heap_size)
+     * allocated by the loader itself will be placed at the end
+     */
+    for (i = stp->num_literals; i-- > 0; )
+    {
```
Nit pick: Move up the { to the previous line.

michalmuskala commented on erts/emulator/beam/ops.tab:
```diff
@@ -1425,7 +1425,11 @@ sorted_put_map_exact F Src=s Dst Live Size Rest=* => \
+   gen_new_map_lit(Dst, Live, Size, Rest)
+
 new_map d I I
+i_new_map_lit d I q
```
I wasn't sure about where this should be placed. It looks like the operations are grouped by release they came to be. On the other hand, it's nice to have all the different map operations together.

bjorng replied:
> On the other hand, it's nice to have all the different map operations together.

I agree.

michalmuskala commented:
I ran the emulator and stdlib test suites and got one failure in [`shell_SUITE:records/1`](https://github.com/erlang/otp/blob/master/lib/stdlib/test/shell_SUITE.erl#L324) - it doesn't look related to me, though.

bjorng commented on erts/emulator/beam/beam_emu.c:
```diff
@@ -7035,6 +7047,66 @@ new_map(Process* p, Eterm* reg, BeamInstr* I)
+    *n_exp = n;
+    ptr = &Arg(4);
+
+    if (n > MAP_SMALL_MAP_LIMIT) {
```
Why is this test done at run-time?

It seems to me that it would be to do this test in the loader and only let this instruction handle small maps.

bjorng commented:
Could you run some more tests before we add the branch to our daily builds?

The test suite for `compiler` runs in a few minutes, but can find a lot of bugs.

The debug-compiled emulator is also good for finding bugs for changes like this. I would recommend running at least the `compiler` and `emulator` tests with a debug-compiler emulator. Running the `stdlib` tests as well would be good.

Before compiling the debug-enabled emulator, you should make a temporary change to erl_map.h like this:

```
@@ -70,11 +70,7 @@ typedef struct flatmap_s {
 #define flatmap_get_keys(x)          (((Eterm *)tuple_val(((flatmap_t *)(x))->keys)) + 1)
 #define flatmap_get_size(x)          (((flatmap_t*)(x))->size)
 
-#ifdef DEBUG
-#define MAP_SMALL_MAP_LIMIT    (3)
-#else
 #define MAP_SMALL_MAP_LIMIT    (32)
-#endif
 
 struct ErtsWStack_;
 struct ErtsEStack_;

```
Without this change, very few small maps will be created.

michalmuskala commented:
Sure, I'll try that and get back with the results.

michalmuskala commented:
@bjorng One more thing I noticed - some places in the loader use `ERTS_ALC_T_LOADER_TMP` and some `ERTS_ALC_T_TMP` for allocating memory - is there a difference?

bjorng commented:
> One more thing I noticed - some places in the loader use ERTS_ALC_T_LOADER_TMP and some ERTS_ALC_T_TMP for allocating memory - is there a difference?

Not really. Use ERTS_ALC_T_LOADER_TMP for new code.

> And another similar one - memcpy vs sys_memcpy

sys_memcpy in new code.

(If you want do cleanups to use ERTS_ALC_T_LOADER_TMP and sys_memcpy in beam_load.c, please do them in separate commits.)

michalmuskala commented:
Running the emulator tests on the debug emulator, the tests got stuck in [`process_SUITE:process_default_tab/1`](https://github.com/erlang/otp/blob/master/erts/emulator/test/process_SUITE.erl#L1528). I killed it after waiting about 15 minutes.

garazdawi commented:
@michalmuskala wait longer, debug tests run takes a long time (5h+) to run. That specific test case takes about 30 min to run.

Alternatively skip scheduler_SUITE:equal_and_high_with_part_time_max, process_SUITE:processes_default_tab, process_SUITE:processes_this_tab which are the test cases that take the longest time to run.

michalmuskala commented:
Both compiler & stdlib suites passed without issues on the debug emulator. I'll rerun the emulator suite, but during the first run, the `map_SUITE` and `map_no_opt_SUITE` passed without issues.

michalmuskala commented:
I finished running the emulator suite and got two failures: `scheduler_SUITE:equal_with_high_max` with reason: `{test_case_failed,{bad,low_min_ratio,0.0}}` and `process_SUITE:processes_default_tab` with reason: `{test_case_failed,{badrpc,nodedown}}`. 

I'm not sure how relevant those are.

josevalim commented:
Should we also consider applying this optimization to copy.c and make sure we share the map literal tuple also when sending messages?

garazdawi commented:
@josevalim that should already work.

garazdawi commented:
beam/beam_load.c:4282:gen_new_small_map_lit() Assertion failed: (((*thp) & 0x3F) == ((0x0 << 4) | ((0x2 << 2) | 0x3)))

This assert is triggered in our debug builds. From what I can see the assert should be moved one line up?

```
(gdb) bt
#0  0x00d77416 in __kernel_vsyscall ()
#1  0x0052e941 in raise (sig=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:64
#2  0x00531e42 in abort () at abort.c:92
#3  0x08273d21 in erl_assert_error (
    expr=0x8358314 "(((*thp) & 0x3F) == ((0x0 << 4) | ((0x2 << 2) | 0x3)))", 
    func=0x8358abd "gen_new_small_map_lit", file=0x8357258 "beam/beam_load.c", 
    line=4282) at sys/unix/sys.c:1291
#4  0x080a9424 in gen_new_small_map_lit (stp=0xb6e405b8, Dst=..., Live=..., 
    Size=..., Rest=0xb7680b08) at beam/beam_load.c:4282
#5  0x080ac358 in transform_engine (st=0xb6e405b8)
    at i686-pc-linux-gnu/debug/smp/beam_tr_funcs.h:32
#6  0x080a2c18 in load_code (stp=0xb6e405b8) at beam/beam_load.c:2149
#7  0x0809e0e5 in erts_prepare_loading (magic=0xb6e40590, c_p=0xb6b00e70, 
    group_leader=8963, modp=0xb6fc0080, code=0xb3cd04f0 "FOR1", 
    unloaded_size=223404) at beam/beam_load.c:751
#8  0x080af643 in prepare_loading_2 (A__p=0xb6b00e70, BIF__ARGS=0xb6fc0080, 
    A__I=0xb5d2729c) at beam/beam_bif_load.c:226
#9  0x0808947d in process_main (x_reg_array=0xb6fc0080, f_reg_array=0xb6fc1100)
    at beam/beam_emu.c:2896
#10 0x081808f4 in sched_thread_func (vesdp=0xb5bc14c0)
    at beam/erl_process.c:8859
#11 0x08316bdb in thr_wrapper (vtwd=0xbfc8c45c) at pthread/ethread.c:118
#12 0x006cccc9 in start_thread (arg=0xb51beb70) at pthread_create.c:304
#13 0x005d469e in clone () at ../sysdeps/unix/sysv/linux/i386/clone.S:130
(gdb) f 4
#4  0x080a9424 in gen_new_small_map_lit (stp=0xb6e405b8, Dst=..., Live=..., 
    Size=..., Rest=0xb7680b08) at beam/beam_load.c:4282
4282	            ASSERT(is_atom(*thp));
(gdb) etp *thp
#Port<0.179346159>.
(gdb) etp thp[-1]
k.
```

bjorng commented:
When you have fixed the faulty assertion, please squash the commits. We will run the tests one more night,  and merge tomorrow unless we'll find any more issues.

michalmuskala commented:
I did notice the assert failure when I was running on the debug emulator and I fixed it locally, but forgot to push. Sorry about that.

I also changed the code to use `sys_memcpy` and the `ERTS_ALC_T_LOADER_TMP` allocator tyoe.

All commits are squashed.

bjorng commented:
When we did our final review of the code, we noticed that the `for` loop in `find_literal()` is written in a confusing way. We have rewritten it in a more conventional way.

We also changed the return value for `find_literal()` to `int`, since that is the type use for functions returning a boolean.

We have pushed a fixup commit. If you agree with out changes, please squash it.

michalmuskala commented:
@bjorng squashed
